neps/nep-519-yield-execution.md

```diff
@@ -0,0 +1,183 @@
+/// replying again.
+///
+/// `method_name_len` and `method_name_ptr`: Identify the callback method that
+/// should be executed either after the contract calls `promise_yield_resume()`
+/// or after 200 blocks have been executed.
+///
+/// `arguments_len` and `arguments_ptr` provide an initial blob of arguments
+/// that will be passed to the callback.  These will be available via the
+/// `input` host function.
+///
+/// `gas`: Similar to the `gas` parameter in
+/// [promise_create](https://github.com/near/nearcore/blob/a908de36ab6f75eb130447a5788007e26d05f93e/runtime/near-vm-runner/src/logic/logic.rs#L1281),
+/// the `gas` parameter is a prepayment for the gas that would be used to
+/// execute the callback.
+///
+/// `gas_weight`: Similar to the `gas_weight` parameter in
+/// [promise_batch_action_function_call_weight](https://github.com/near/nearcore/blob/a908de36ab6f75eb130447a5788007e26d05f93e/runtime/near-vm-runner/src/logic/logic.rs#L1699),
+/// this improves the devX for the smart contract.  It allows a contract to
+/// specify a portion of the remaining gas for executing the callback instead of
+/// specifying a precise amount.
+///
+/// `register_id`: is used to identify the register that will be filled with a
+/// unique resumption token. This token is used with `promise_yield_resume` to
+/// resolve the continuation receipt set up by this function.
+///
+/// Return value: u64: Similar to the
+/// [promise_create](https://github.com/near/nearcore/blob/a908de36ab6f75eb130447a5788007e26d05f93e/runtime/near-vm-runner/src/logic/logic.rs#L1281)
+/// host function, this function also create a promise and returns an index to
+/// the promise.  This index can be used to create a chain of promises.
+pub fn promise_yield_create(
+    method_name_len: u64,
+    method_name_ptr: u64,
+    arguments_len: u64,
+    arguments_ptr: u64,
+    gas: u64,
+    gas_weight: u64,
+    register_id: u64,
+) -> u64;
+
+/// See `promise_yield_create()` for more details.  This host function can be
+/// used to resolve the continuation that was set up by
+/// `promise_yield_create()`.  The contract calling this function must be the
+/// same contract that called `promise_yield_create()` earlier.  This host
+/// function cannot be called for the same resumption token twice or if the
+/// callback specified in `promise_yield_create()` has already executed.
+///
+/// `data_id_len` and `data_it_ptr`: Used to pass the unique resumption token
+/// that was returned to the smart contract in the `promise_yield_create()`
+/// function.
+///
+/// `payload_len` and `payload_ptr`: the smart contract can provide an
+/// additional optional blob of arguments that should be passed to the callback
+/// that will be resumed.  These are available via the `promise_result` host
+/// function.
+pub fn promise_yield_resume(
+    data_id_len: u64,
+    data_id_ptr: u64,
+    payload_len: u64,
+    payload_ptr: u64,
+) -> ();
```

Comment:
One more small change that has happened to the implementation since the last review:

```suggestion
/// See `promise_yield_create()` for more details.  This host function can be
/// used to resolve the continuation that was set up by
/// `promise_yield_create()`.  The contract calling this function must be the
/// same contract that called `promise_yield_create()` earlier.  This host
/// function cannot be called for the same resumption token twice or if the
/// callback specified in `promise_yield_create()` has already executed.
///
/// `data_id_len` and `data_it_ptr`: Used to pass the unique resumption token
/// that was returned to the smart contract in the `promise_yield_create()`
/// function.
///
/// `payload_len` and `payload_ptr`: the smart contract can provide an
/// additional optional blob of arguments that should be passed to the callback
/// that will be resumed.  These are available via the `promise_result` host
/// function.
///
/// The function returns `1` if submitting the payload was successful. This 
/// guarantees that the yielded callback will be executes with one of the
/// successfully submitted payloads as the input. Otherwise (e.g. if the yield 
/// receipt has already timed out) `0` will be returned, indicating that this
/// payload could not be submitted successfully.
pub fn promise_yield_resume(
    data_id_len: u64,
    data_id_ptr: u64,
    payload_len: u64,
    payload_ptr: u64,
) -> u32;
```

> This host function cannot be called for the same resumption token twice or if the callback specified in `promise_yield_create()` has already executed.

We may also want to relax this requirement. I.e. allow returning `true` more than exactly once. In that scenario, the callback is allowed to execute with any of the payloads that have been supplied and for which `true` was returned.

cc @saketh-are for double-check.

Reply:
+1, if we can relax this it will make things simpler for us. The guarantees we can more easily provide are:
- `promise_yield_resume` will return true iff the callback has not yet been executed.
- If the user gets back true from `promise_yield_resume`, the callback will execute with a user-provided payload rather than a timeout signal.
- If the user calls resume multiple times and gets back multiple trues, the callback will be executed with one of the given payloads and the others will be safely ignored.

Hopefully we can come up with a simple and clear way to condense all of this into the docs.